src/doc-utils.js

```diff
@@ -1,38 +1,48 @@
 "use strict";
-function iterDoc(topDoc, func) {
-  const docs = [ topDoc ];
-  while (docs.length !== 0) {
-    const doc = docs.pop();
-    let res = undefined;
-
-    if (typeof doc === "string") {
-      const res = func("string", doc);
-
-      if (res) {
-        return res;
-      }
-    } else {
-      const res = func(doc.type, doc);
-
-      if (res) {
-        return res;
-      }
+function traverseDocUpDown(doc, onEnter, onExit) {
+  if (doc.type === "concat") {
+    onEnter(doc);
+    for (var i = doc.parts.length - 1; i >= 0; i--) {
+      traverseDocUpDown(doc.parts[i], onEnter, onExit);
+    }
+    onExit(doc);
+  } else if (doc.type === "if-break") {
+    onEnter(doc);
+    if (doc.breakContents) {
+      traverseDocUpDown(doc.breakContents, onEnter, onExit);
+    }
+    if (doc.flatContents) {
+      traverseDocUpDown(doc.flatContents, onEnter, onExit);
+    }
+    onExit(doc);
+  } else if (doc.contents) {
+    onEnter(doc);
+    traverseDocUpDown(doc.contents, onEnter, onExit);
+    onExit(doc);
+  } else {
+    onEnter(doc);
```

Comment:
may want to also onExit here. Otherwise if you do counts based on onEnter/onExit it's not going to sum to 0.